packages/node/src/indexer/fetch.service.ts

```diff
@@ -45,6 +58,107 @@ export class FetchService implements OnApplicationShutdown {
     return this.apiService.getApi();
   }
 
+  getIndexFilters(): ProjectIndexFilters {
+    const dataSources = this.project.dataSources.filter(
+      (ds) =>
+        !ds.filter?.specName ||
+        ds.filter.specName === this.api.runtimeVersion.specName.toString(),
+    );
+    const response = {
+      existBlockHandler: false,
+      existEventHandler: false,
+      existExtrinsicHandler: false,
+      eventFilters: [],
+      extrinsicFilters: [],
+    } as ProjectIndexFilters;
+    for (const ds of dataSources) {
+      if (ds.kind === SubqlKind.Runtime) {
+        for (const handler of ds.mapping.handlers) {
+          switch (handler.kind) {
+            case SubqlKind.BlockHandler:
+              response.existBlockHandler = true;
+              break;
+            case SubqlKind.CallHandler: {
+              response.existExtrinsicHandler = true;
+              if (
+                response.extrinsicFilters.findIndex(
+                  (extrinsic) =>
+                    extrinsic.module === handler.filter.module &&
+                    extrinsic.method === handler.filter.method,
+                ) <= -1
+              ) {
+                response.extrinsicFilters.push(handler.filter);
+              }
+              break;
+            }
+            case SubqlKind.EventHandler: {
+              response.existEventHandler = true;
+              if (
+                response.eventFilters.findIndex(
+                  (event) =>
+                    event.module === handler.filter.module &&
+                    event.method === handler.filter.method,
+                ) <= -1
+              ) {
+                response.eventFilters.push(handler.filter);
+              }
+              break;
+            }
+            default:
+          }
+        }
+      }
+    }
+    return response;
+  }
+
+  isUseDictionary() {
+    if (!this.project.network.dictionary) {
+      return false;
+    }
+    if (
+      this.projectIndexFilters.existBlockHandler ||
+      (this.projectIndexFilters.existEventHandler &&
+        this.projectIndexFilters.eventFilters.length === 0) ||
+      (this.projectIndexFilters.existExtrinsicHandler &&
+        this.projectIndexFilters.extrinsicFilters.length === 0)
+    ) {
+      return false;
+    }
+
+    return true;
+  }
+
+  async getDictionary(startBlock: number) {
+    const indexFilters = this.projectIndexFilters;
+    if (
+      indexFilters.existBlockHandler ||
+      (indexFilters.existEventHandler &&
+        indexFilters.eventFilters.length === 0) ||
+      (indexFilters.existExtrinsicHandler &&
+        indexFilters.extrinsicFilters.length === 0)
+    ) {
+      this.useDictionary = false;
+    }
+    const query = this.dictionaryService.dictionaryQuery(
```

Comment:
`dictionaryService.dictionaryQuery` doesn't have to be exported. 
or let's say `dictionaryService.dictionaryQuery`, `this.dictionaryService.getDictionary`, and `this.dictionaryService.getBlockBatch` can be merged into one method `this.dictionaryService.getBlockBatch(startBlock, batchSize,...handlers)` we don't really need these middle results